all item bag menu

SamKacer opened the pull request:
Note that there was a mix of spaces and tabs used for indentation in LocalMenu.lua, so I reindented it to use tabs only for consistency. Due to this, there is a ton of line changes that are just indentation changes. To filter these out when looking at the diff, please make sure to tick the "hide whitespace" option in diff options.

## Functional changes
- "all items" menu was added to bags, which lists all items alphabetically, with newly acquired items on top

## Refactoring changes
- deleted some unused vars (bunch of tText and tTextFull)
- replaced some undefined vars with nil
- factored out some code duplication
- made some variable names more specific (e.g. renamed tFriendlyName in different instances)

The last point I was doing when I was initially trying to understand the code. The same var name being used in different contexts (tFriendlyName specifically) made it hard to keep in mind to what it was referring to in a given context.
Renaming vars called "tFriendlyName" to something more specific for its given context made the code easier to read. Since, this might aid other devs in the future as well, I kept that refactoring there.

## Known bugs
- for some reason, "new" is prepended twice to the beginning of new items

## Remaining effort
- german translation for "all items"

#### commit messages

- Extract out containerFrame var
- Reindent LocalMenu.lua with only tabs
- Rename tFriendlyName to bagName
- Rename tFriendlyName to bagItemSlotName
- Simplify appending bag result
- Implement all items menu
- Filter out empty from all items view
- Sort all items view alphabetically
- Remove manual numbering in all items bag view
- Remove unused or undefined variables
- Clarify bag result variable names
- Show newly acquired items first
- Fix items like "empty vial" being filtered out
- Add "all items" to locale strings

Duugu commented:
Hey, thanks a lot!

tFriendlyName was intentionally general, because that variable is holding the displayed name.
I am re-using that code to manually implement more game windows later on.
Removing unused variables, changing variable names to specific terms like "bags" and replacing nil variables by nil makes that more effort.
I would prefer if you could keep my code as it is. At least in the local module. Thank you! :)

I guess that you are aware, but just to be 100% sure... please be careful with replacing table[#table +1] by table.insert. 
The results are not always the same. Especially with indexed tables with holes.
And table.insert actually is much slower. That isn't an issue here. But please keep that in mind if in other parts of the code. In big tables it can lead to very, very slow code.
I would suggest to keep existing implementions, just to avoid unintended side effects. At least if a change just is for "better readability".
Is you are 100% sure that changing that to table.insert somewhere has no side effects, then please feel free to do so, though.

SamKacer commented:
@Duugu thanks for feedback!

I went ahead and changed the table.insert back to what it was, since I didn't think of how it might behave differently with tables with holes. Thanks for the headsup :)

I changed back `bagItemSlotName` to `tFriendlyName`, since I realized the main problem with code duplication and lack of readability was due to all the `aParentsChilds[tFriendlyName]`, so since I extracted those out into a more specifically named variable, `tFriendlyName` was only used in one spot and so didn't really affect readability.

I'm hoping I can change your mind about the removal of unused and undefined variables. There is a very good reason every code linter out there flags unused variables as a warning. In the best scenario they just make things harder to read, since there is extra stuff that does nothing. But in the worst scenario they are a source of bugs, since they usually signify the developer defined it and forgot to use it. In Go lang unused variables are even treated as errors and the program won't compile if it has them.

Use of undefined variables are even worse! Since these are *always* a bug and even more misleading when reading the code.

But, of course I understand the concern about being able to reuse the code. However, here I am inferring that by reusing the code, you mean having it be easy to copy and paste, which is actually not what devs generally mean when they talk about reusing code for very good reason :)

For one, copy pasted code makes things way harder to read, especially for other devs, but even for yourself down the line. Whenever, there are very similar blocks of code sprinkled throughout a codebase, a dev reading has to carefully compare between them to find what the actual differences are.

But what is much worse, is excessive copy pasting leads to calcification of the codebase. Imagine that you thought of a new UI feature and for it to work you required to add a new attribute to all those button tables. There are so many instances that the effort to change it in all those different places would be so much effort that it wouldn't even be worth it to do. Imagine if even a more complex change was necessary.

I am wondring, why not have a function for those initial couple of lines that are copy pasted when defining local menus? That way the function could have sensible defaults, since in most instances some things like type, RoC, childs, obj, and others have thesame value. Also, this way if there ever needed to be a change to all menus, the change would need to be only made in one or two places, instead of where all menus were defined. And even if the change would need to made at the call site, hunting them all down would be much easier since the IDE can list all references to a function/variable. Highly recommend the lua language server extension for VS code if not already using! :)

Also there wouldn't even be a need for these local variables tFriendlyName, tText, and tTextFull, making subsequent code more readable .

What I mean is something like this (better names may be appropriate):

```
local function createChild(childs, friendlyName, attributes)
    local newChild = {
		frameName = attributes.frameName,
		RoC = attributes.RoC or "Child",
		type = attributes.type or "Button",
		obj =attributes.obj or  _G[attributes.frameName],
		textFirstLine = attributes.textFirstLine or friendlyName,
		textFull = attributes.textFull oor "",
		noMenuNumbers = attributes.noMenuNumbers or true,
		childs = attributes.childs or {},
	}
	table.insert(childs, newChild)
	childs[friendlyName] = newChild
	return newChild
end
```

Then the definition of the bags menu could be defined like this:

```
--BEFORE
local tFriendlyName = L["Bags"]
   local tText, tFullText = L["Bags"], ""
   table.insert(aParentChilds, tFriendlyName)
   aParentChilds[tFriendlyName] = {
      frameName = tFrameName,
      RoC = "Child",
      type = "Button",
      obj = _G[tFrameName],
      textFirstLine = tFriendlyName,
      textFull = "",
      noMenuNumbers = true,
      childs = {},
      func = nil,
      click = true,
   

--AFTER
local bagsMenu = createChild(aParentsChilds, L["Bags"], {})
bagsMenu.click = true
```

I'll give a couple more examples to give more of n idea of what the difference would be

```
--BEFORE
				local tFriendlyName = L["Bag"].." "..bagId
				local tText, tFullText = L["Bag"].." "..bagId, ""
				table.insert(aParentChilds, tFriendlyName)
				aParentChilds[tFriendlyName] = {
					frameName = tFrameName,
					RoC = "Child",
					type = "Button",
					obj = _G[tFrameName],
					textFirstLine = tFriendlyName,
					textFull = "",
					noMenuNumbers = true,
					childs = {},
				}   

--AFTER
local bagMenu = createChilds(aParentsChilds, L["Bag"].." "..bagId, {frameName = tFrameName})

--BEFORE
local tFrameName = "TradeSkillFrame"
local tFriendlyName = _G["TradeSkillFrameTitleText"]:GetText()
table.insert(aParentChilds, tFriendlyName)
aParentChilds[tFriendlyName] = {
	frameName = tFrameName,
	RoC = "Child",
	type = "FontString",
	obj = _G[tFrameName],
	textFirstLine = tFriendlyName,
	textFull = "",
	childs = {},
}

--AFTER
local tFrameName = "TradeSkillFrame"
local tradeSkillMenu = createChild(aParentsChilds, _G["TradeSkillFrameTitleText"]:GetText(), {
	frameName= tFrameName,
	type = "FontString",
})
```

It wouldn't necessarily have to  be done exactly this way. But I am just trying to show that there are ways to handle code reuse without having to have unused or undefined variables, with even additional benefits like code being more concise, and making modifications easier.
